### ISSUE TITLE
Pointer Masking ACT

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#15]
### Technical Group

Applications & Tools HC

### ratification-pkg

Pointer Masking

### Technical Liaison

Martin Maas, Adam Zabrocki

### Task Category

Arch Tests

### Task Sub Category

- [ ] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

SOW: [link](https://docs.google.com/document/d/1BPZ2PK2RQWZxfaofNcDTGjG1t0ZIgMVXd4AHlMcPw_c/)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - ACT SIG sign-off date (if ACT work):

### Waiver

- [ ] Freeze
- [ ] Ratification

**Comment by jjscheel** [comment]
@mustafa7755, @Muhammad-Saadi-Abbasi, it sounds like you guys will be working on this.  I've assigned it to both of you not knowing if you're sharing or working on separate pieces.  Feel free to unassign yourselves as-appropriate based on assignments.

**Comment by UmerShahidengr** [comment]
The planning phase of this project has been completed, this is the test plan for the viewers to review: [Test plan Pointer Masking Ext](https://docs.google.com/spreadsheets/d/1YzIjz79Zrf10NepTPnQYcmTFpYiTuG-2cca8Q-70BGk/edit?usp=sharing)
Goal for next week: Develop the skeleton code (Basic R/W access to the Pointer masking CSR regs) according to the ACT template, and run it on Spike via riscof.

**Comment by UmerShahidengr** [comment]
@martinmaas, @Adam-pi3 Sir, Please find some time to review the test plan. You are the domain experts, so your comments would be valuable.

**Comment by martinmaas** [comment]
Apologies for the delayed response (I was on vacation and it took me a while to catch up on everything after coming back). Thank you so much for working on this!

My main feedback is that the latest version of the pointer masking spec has removed instruction pointer masking, which obviates the need for tests related to instructions and to the *ipmen fields.

https://github.com/riscv/riscv-j-extension/blob/master/zjpm-spec.pdf

Some additional feedback:
* We will still need to test interrupts/exceptions and how pointer masking behaves when changing between different privilege levels.
* All cases need to be tested for all privilege modes and address translation modes (Sv39, Sv48, Sv57 and Bare) that are supported.
* There is an additional corner case where a misaligned address wraps around (e.g., reading 64 bit from 0x...fffffa). We should cover this in the tests.
* We should make to ensure that we cover all instructions subject to pointer masking. In addition to the ones on the second page, this now includes CMOs as well.
* We should test that the behavior for MPRV=1 is correct.
* Pointer masking is not currently defined for RV32, so no testing is needed for that case.

Please don't hesitate to let @Adam-pi3 and me know if there are any additional questions.

**Comment by UmerShahidengr** [comment]
Update May 7th, 2023 ⇾ Not much update on this task. Possibly it will get deferred to Q3, 2023 because of upcoming new work (debug native triggers)

**Comment by jjscheel** [comment]
@martinmaas, thanks for the update.  Would you kindly ensure that your additional comments about ACT get into the referenced SOW document ([link](https://docs.google.com/document/d/1BPZ2PK2RQWZxfaofNcDTGjG1t0ZIgMVXd4AHlMcPw_c/))?

**Comment by jjscheel** [comment]
Thanks for the update, @UmerShahidengr.  I recognize we're in a holding pattern here.

I'm going to leave on the Agenda for discussion purpsoses at this time.

@martinmaas, @Adam-pi3, you folks should be aware that given the challenges we've had with Code Size Reduction, I've got 1 resource to spend on either Pointer Masking and/or Debug.  So, barring some explicit priority guidance, they will be started first-come, first serve.  I'll be in touch with what and how we decide.

**Comment by jjscheel** [comment]
Correction, the challenges with Pointer Masking ACT have been in Zfinx, Zfh, Zdinx, and Zhinx.

**Comment by jjscheel** [comment]
The discussion this week in the Pointer Masking TG was that @martinmaas would request the Architecture Review Committee to prioritize closure of the items preventing SAIL and ACT work in hopes of keeping this work moving forward.

@martinmaas, please provide an update when available.

**Comment by UmerShahidengr** [comment]
Update May 23rd, 2023 => Not much update on this task yet.

**Comment by Muhammad-Saadi-Abbasi** [comment]
@jjscheel, the tool-chain doesn't has support for `Zjpm` extension.

**Comment by martinmaas** [comment]
Pointer masking does not add any new instructions and can thus be used without toolchain support. Once the spec has passed architecture review, the toolchain will need to add some constants, but those shouldn't block ACT or Sail.

An (out of date) example can be found here: https://github.com/martinmaas/riscv-arch-test/commit/f3ba7f8c65496f46a11409b938888a6cb5cb1a56

Please also note that the latest changes requested by the architecture review committee will remove the new CSRs that pointer masking is adding and use the existing *envcfg CSRs.

**Comment by jjscheel** [comment]
Thanks, @martinmaas!  Please keep us posted on the the ARC changes.

**Comment by jjscheel** [comment]
Agreed.  Removing from Agenda.  Re-visit next meeting.

**Comment by jjscheel** [comment]
Moving Projected completion out 6 months based on where we are to 12/30.

**Comment by UmerShahidengr** [comment]
Update ⇾ July 11th, 2023
Blocked

**Comment by jjscheel** [comment]
Per ARC minutes from 8/23/2023: https://lists.riscv.org/g/tech-chairs/message/1593

> The ARC continued discussion of the Pointer Masking spec, with the primary issue being the extension and enable bit names. The previous suggestion that the enable bit names be tied to the extension names was deemed inappropriate upon reflection because Pointer Masking an attribute of the ABI for given mode of operation, and the extension names are for the bits at higher modes of operation that define the enable bits. It was also suggested that the number of address bits be included in the name (not the number of ignored bits to make it independent of RV64 vs. RV128). As a result, our recommendation is now that the Pointer Masking enable bits be named PM48E in their respective CSRs. This enables would be read-only 0 in RV32 modes of operation.

**Comment by UmerShahidengr** [comment]
Update ⇾ September 26th, 2023
No update on this.

**Comment by UmerShahidengr** [comment]
Update ⇾ October 24th, 2023
No update on this.

**Comment by jjscheel** [comment]
From ARC Minutes for 11/14/2023: https://lists.riscv.org/g/tech-unprivileged/message/585

> An updated pointer-masking spec was received, and ARC expects to complete final review and approve next week.

Unblocking.  @UmerShahidengr, as they say in racing, "Start your engines!"

**Comment by UmerShahidengr** [comment]
Update ⇾ November 28th, 2023
Great to hear that. Thanks for the update @jjscheel

**Comment by UmerShahidengr** [comment]
Update ⇾ December 12th, 2023
@HAMZA-AFZAL-404 will work on this in parallel to the Sail implementation

**Comment by jjscheel** [comment]
Thanks, @UmerShahidengr!  Please ensure that Hamza gets a RISC-V Portal id, joins the devpartners community, and places their GH id in their profile.  Questions can come directly to me.

**Comment by UmerShahidengr** [comment]
Update ⇾ March 5th, 2024
Stalled as there is no compiler support for Pointer Masking, so we cant write any assembly tests to verify it.

**Comment by jjscheel** [comment]
According to the [Pointer Masking ISA Status Checklist, "Freeze" tab](https://docs.google.com/spreadsheets/d/1OP81ZVp5AClSKK9CzexW6Kuvb3NQG7JlH5sGRMT5QQc/edit#gid=1983095369) only CSR access is needed for Pointer Masking support.  

@martinmaas, @Adam-pi3, can you confirm that his is the case and assist with an explanation of how test should be written?

**Comment by martinmaas** [comment]
That's correct – pointer masking doesn't add any additional CSRs or instructions (it only defines new bits in existing CSRs). No compiler support is needed.

**Comment by UmerShahidengr** [comment]
Pointer Masking defines new bits in existing CSRs, but those bits were reserved earlier. Accessing those reserved bits are giving compile time errors.
@HAMZA-AFZAL404 , @muhammad-maarij-zeeshan can you share your insights here.

**Comment by martinmaas** [comment]
Which compiler are you using, and what is the error message you are seeing? I had written some simple one-off tests (using gcc) and they compiled without problems, but it could be different for LLVM or other versions of GCC. If this does require changes to the compiler, it is probably a very simple patch.

**Comment by jjscheel** [comment]
Thanks, @martinmaas.  @UmerShahidengr, @HAMZA-AFZAL404 , or @muhammad-maarij-zeeshan can you provide some details, please?

**Comment by jjscheel** [comment]
@martinmaas, @Adam-pi3, your guidance on @HAMZA-AFZAL404 's post would be appreciated.  Thanks!

**Comment by martinmaas** [comment]
For physical addresses, pointer masking with 16 bits means that the effective address is the same as if the top bits are set to 0.

Looking at the remaining part of the spreadsheet: Sv48 doesn't necessarily have to result in a page fault – in that case, the effective address is the same as if the top 16 bits are the sign extension of bit 47.

**Comment by HAMZA-AFZAL404** [comment]
Thank you, Martin, for your insights. For more clarification i want to discuss a  specific scenario related to pointer masking on physical addresses.**Correct me if i am wrong please**

Let’s consider an example where the original address is `0x0080000000000688`. If we apply pointer masking with PMLEN = 16, the resulting address becomes `0x0000000000000688`.

So Far as i understand address probably gona be invalid. So it will lead to an access fault.

**Comment by martinmaas** [comment]
For a physical address, the resulting address with PMLEN=16 becomes `0x0080000000000688`. For a virtual address, it becomes `0xFF80000000000688`. Whether or not it becomes an access fault depends on whether this part of the address space is valid/mapped.

**Comment by UmerShahidengr** [comment]
Update ⇾ April 2nd, 2024
@HAMZA-AFZAL404 has been working on ACTs. The work is in development phase

**Comment by HAMZA-AFZAL404** [comment]
> For a physical address, the resulting address with PMLEN=16 becomes `0x0080000000000688`. For a virtual address, it becomes `0xFF80000000000688`. Whether or not it becomes an access fault depends on whether this part of the address space is valid/mapped.

Lets say any kind of fault happens on accessing this address what should be wrriten in the tval .The "masked address" or the "unmasked address".

**Comment by martinmaas** [comment]
Apologies for the delayed response. Section 2.6 describes the details of when the masked or unmasked address should be used. For hardware writes to registers such as tval, pointer masking *is* applied (i.e., the tag is removed).

**Comment by UmerShahidengr** [comment]
Update ⇾ April 18th, 2024
ACTs are under development. Almost all tests are ready except pointer masking with sv57 and pointer masking with vector operations. 
For the time being, we are ignoring ACTs related to pointer masking with vector operations, rest of the tests will be delivered before our next meeting.
CC: @HAMZA-AFZAL404

**Comment by UmerShahidengr** [comment]
Update April 30th, 2024:
ACTs are ready but the coverpoint definitions are yet to be added,

**Comment by jjscheel** [comment]
@UmerShahidengr, what is the status of the coverpoints definitions?  Any progress?

**Comment by martinmaas** [comment]
Thanks @UmerShahidengr for the work on this! Is there a PR for the ACTs, even if the coverpoint definitions are missing?

**Comment by martinmaas** [comment]
(My reason for asking is that we need to validate the Spike port, and even incomplete ACTs would be very helpful.)

CC: @HAMZA-AFZAL404 @UmerShahidengr

**Comment by jjscheel** [comment]
@UmerShahidengr, any update here?

**Comment by UmerShahidengr** [comment]
Update June 11th, 2024:
The ACTs are available [here](https://github.com/HAMZA-AFZAL404/riscv-arch-test/tree/pmm_tests/riscv-test-suite/rv64i_m/pmm), its PR has not been added yet as its coverpoint definitions are still missing.

**Comment by jjscheel** [comment]
Thanks, @UmerShahidengr.  When will the coverpoint definitions be drafted and ready for submission with PR?

**Comment by UmerShahidengr** [comment]
@jjscheel we will target to complete it in the next month.

**Comment by jjscheel** [comment]
Coverage information from @Adam-pi3:

**Definitions**

- Aligned access - any reference to "aligned access" includes all possible variations:
  - 8-byte reads must be 8-byte aligned
  - 4-byte reads must be 4-byte aligned
  - 2-byte reads must be 2-byte aligned
- Non-aligned access - any reference to "non-aligned access" includes all possible variations:
  - 8-byte reads must NOT be 8-byte aligned
  - 4-byte reads must NOT be 4-byte aligned
  - 2-byte reads must NOT be 2-byte aligned
- RISC-V mode - any reference to "mode" includes all implemented and supported NRISC-V modes, which might include:
  - HS/S-mode
  - VS-mode
  - (V)U-mode
- PM Mask - references various possible
  - no mask
  - 7 bits mask
  - 16 bits mask
- "PM enabled" - enabling PM
  - any reference to "PM enabled" includes enabling PM and testing all variation of "PM Mask"
- "PM disabled" - disabling PM in CSR register

**Based on those definitions, we proposed the following test plans (“coverpoints”):**

PM works on Physical Address (PA) or Virtual Address (VA) but not on both at the same time. If current mode currently leverages Virtual Memory feature, PM is masking VA. Otherwise PM is enforced on PA. For example, if U-mode code is running under Virtual Memory controlled by S-mode, PM for S and U mode is masking VA. However, if at some point CPU switches to M-mode which operates on physical memory, PM on M-mode would be enforce on PA. **The following tests should be executed for PA and VA independently.**

 

1. Standard LOAD / STORE execution and validate expected results (e.g., read value or generated fault), during:
    1.  Normal execution     
        1. With "PM enabled" in all "RISC-V modes"   [should cover PM function cross {all_lsu_instruction, [bare_mode, nvmpu, hash_mpn], [mmode, smode, umode]} ]
        1. With "PM disabled" in all "RISC-V modes"
    1. Execution in exception context - if normal execution generates an exception and continue execution logic in exception context, verify if PM state is NOT modified.
        1. With "PM enabled" in all "RISC-V modes"
        1. With "PM disabled" in all "RISC-V modes"
    1. Execution in interrupt context - if normal execution is interrupted and continue execution logic in interrupt context, verify if PM state is NOT modified.
        1. With "PM enabled" in all "RISC-V modes"
        1. With "PM disabled" in all "RISC-V modes"
    1. Execution in NMI (with disabled IRQs) - if normal execution is interrupted by NMI with IRQ disabled and continue execution logic in NMI context, verify if PM state is NOT modified.
        1. With "PM enabled" in all "NVRISC-V modes" 
        1. With "PM disabled" in all "NVRISC-V modes"
1. Execute LOAD / STORE and validate expected results (e.g., read value or generated fault), when referencing:
    1. "Aligned access"
        1. With "PM enabled" in all "RISC-V modes"
        1. With "PM disabled" in all "NVRISC-V modes"
    1. "Non-aligned access"
        1. With "PM enabled" in all "RISC-V modes"
        1.  With "PM disabled" in all "RISC-V modes"
    1. "Edge" of the page
        1. Crossing page boundary
            1. With "PM enabled" in all "RISC-V modes"
            1. With "PM disabled" in all "RISC-V modes"
        1. Not crossing page boundary
            1. With "PM enabled" in all "RISC-V modes"
            1. With "PM disabled" in all "RISC-V modes"
1. Initialization values:
    1. CSR PM related registers
1. Non-FNL SW test plan - Context switch - validate if all PM related CSRs are preserved during context switch when:
    1. All PM related CSRs are enabled in all "RISC-V modes" at the same time
    1. All PM related CSRs are enabled in just one specific mode "RISC-V modes" at the time
    1. <other modes are NV RISC-V specific implementation and does not apply to general PM case> - out of scope

**Comment by jjscheel** [comment]
@UmerShahidengr  ^^

Can you please review and comment if this helps with the coverpoint definitions?

**Comment by UmerShahidengr** [comment]
@jjscheel thank you so much for sharing, I have written some coverpoints based on the proposed testplan. There are some issues in Sail model which is causing some delay.
For example, one problem is to define `PMM` bits in `mseccfg` register, this register is `XLEN` bit register, and `PMM` bits are `33 .. 32`. These bits can only be defined for RV64 (as J-extension is only for RV64, not for RV32). There is no way to tell Sail that `PMM` bitfield is only for RV64 version and not for RV32 version.
If we declare `mseccfg` to be a 64bit register then it does not build for RV32 and other Zicsr tests start failing, otherwise if we declare `mseccfg` to XLEN bit and build Sail for RV32, then it starts giving Syntax error as bit 32,32 are out of bound for RV32.

We have seen one potential solution applied in Sail code [here](https://github.com/riscv/sail-riscv/blob/1711385730a96c18f6b89ba47e5030136b54c0c0/model/riscv_sys_regs.sail#L198) but this ad-hoc has worked for a bitfield name of a single bit (MBE, which does not exist in RV32 but exists in Rv64), since PMM is multi-bit field so managing this case is becoming complicated.

@billmcspadden-riscv can you please comment on how to address this issue?

**Comment by martinmaas** [comment]
Thank you so much @UmerShahidengr! Regarding your question:

* The pointer masking extension does reserve those bits on RV32 bit as well, not just in the presence of the pointer masking extension (the bits are defined in the main RISC-V ISA spec; only their use is governed by the pointer masking extension). See https://github.com/riscv/riscv-isa-manual/pull/1170

* The pointer masking spec defines the behavior of the fields on RV32, which is that they are hard-wired to zero.

* I believe `mseccfg` is not a 64 bit register on RV32 but instead will be accessed through two 32-bit registers, `mseccfg` and `mseccfgh` (similar for the other 64-bit CSRs).

The approach in the Sail code you linked looks correct to me in this case as well (update the field only when XLEN == 64). What prevents this approach from being used for PMM?

**Comment by UmerShahidengr** [comment]
Update July 23rd, 2024:
Thank you @martinmaas for the detailed comment. We used the same approach and it worked. The Sail PR has already been reviewed, now we are proceeding towards the ACTs

**Comment by martinmaas** [comment]
This is great to hear! I just took a look at the Sail PR and it looks great – please let me know if and when I should take a look at the ACTs as well. Thank you once again for all the work on this!

**Comment by jjscheel** [comment]
@martinmaas, if @UmerShahidengr can pull together a RFC PR, that would be easiest to review.  But, yes, I think we're close to verifying all tests a feature complete and coverpoints defined.  @UmerShahidengr, do you agree?

**Comment by jjscheel** [comment]
Ooops, I forgot that @UmerShahidengr reported the tests are only about 70% complete.  So, @martinmaas, we are not yet ready for a review.

However, we could review the coverpoints file if @Adam-pi3 is willing and @UmerShahidengr will share.  That would only leave the testcases when done.

**Comment by jjscheel** [comment]
@UmerShahidengr, can you create an (RFC) PR for the cover points?  They're functionally complete, right?  This would allow Martin and Adam to review and comment.

**Comment by UmerShahidengr** [comment]
Update August 6th, 2024
The ACTs are released as PR. PR can be reviewed [here](https://github.com/riscv-non-isa/riscv-arch-test/pull/484)

**Comment by Adam-pi3** [comment]
I schimmed through the tests and I've been in sync with @UmerShahidengr  regarding the functional problem which @UmerShahidengr  already fixed. Nevertheless, per my understanding as of now the test covers:
 1.i.
 2.i
 3

there is missing for 1.ii/iii/iv, 2.ii/iii/iv and 4. Did I miss them?
Additionally, for the test which we have, I believe we should verify address with and without tag value. For now we only generate addresses without tag value. Example:

 - with PMM disable we test
   -> 0x0000000012345678 (we assume it points to the valid memory) - it should succeed
   -> 0xAA00000012345678 - it should fail and generate exception which we gracefully handle
 - with PMM enabled we test:
   -> 0x0000000012345678 (we assume it points to the valid memory) - it should succeed
   -> 0xAA00000012345678 - it should succeed too and point to the same address and value as previous load.

**Comment by jjscheel** [comment]
@UmerShahidengr, can you respond to @Adam-pi3 review comments?  I'm particularly interested in whether the mentioned gaps exist in the current ACT tests?  Are they included in the cover reports?

**Comment by UmerShahidengr** [comment]
@Adam-pi3 There is the inline reply to each coverpoint required:

### Standard LOAD / STORE Execution and Validation of Expected Results

- Normal Execution
  - With "PM enabled" in all "RISC-V modes" [Should cover PM function across `{all_lsu_instruction, [bare_mode, nvmpu, hash_mpn], [mmode, smode, umode]}`]
  - With "PM disabled" in all "RISC-V modes"
  - **Comment: Done for all cases.**

- **Execution in Exception Context**
  - If normal execution generates an exception and continues execution logic in exception context, verify if PM state is NOT modified.
  - With "PM enabled" in all "RISC-V modes"
  - With "PM disabled" in all "RISC-V modes"
  - **Comment:** It has also been done. It is not checked within the trap handler if PM state is getting modified or not, but after exiting the trap, it is checked in most of the tests if the state persists or not.

- **Execution in Interrupt Context**
  - If normal execution is interrupted and continues execution logic in interrupt context, verify if PM state is NOT modified.
  - With "PM enabled" in all "RISC-V modes"
  - With "PM disabled" in all "RISC-V modes"
  - **Comment:** No test has been written in interrupt context yet, as `riscv-arch-tests` trap handler has not been tested for any interrupt, and Sail has not been tested for basic interrupts yet. So this will be on hold until basic interrupt tests are developed.

- **Execution in NMI (with Disabled IRQs)**
  - If normal execution is interrupted by NMI with IRQ disabled and continues execution logic in NMI context, verify if PM state is NOT modified.
  - With "PM enabled" in all "NVRISC-V modes"
  - With "PM disabled" in all "NVRISC-V modes"
  - **Comment:** There is no support in Sail to generate NMI, so this cannot be verified as of now.

### Execute LOAD / STORE and Validate Expected Results

- **Aligned Access**
  - With "PM enabled" in all "RISC-V modes"
  - With "PM disabled" in all "RISC-V modes"
  - **Comment:** Done.

- **Non-aligned Access**
  - With "PM enabled" in all "RISC-V modes"
  - With "PM disabled" in all "RISC-V modes"
  - **Comment:** Not done yet, but will be released in the next commit.

- **Edge of the Page**
  - Crossing Page Boundary
    - With "PM enabled" in all "RISC-V modes"
    - With "PM disabled" in all "RISC-V modes"
    - **Comment:** Done.
  
  - Not Crossing Page Boundary
    - With "PM enabled" in all "RISC-V modes"
    - With "PM disabled" in all "RISC-V modes"
    - **Comment:** Done.

### Initialization Values
- **CSR PM Related Registers**
  - **Comment:** Done.

### Non-FNL SW Test Plan - Context Switch

- Validate if all PM related CSRs are preserved during context switch when:
  - All PM related CSRs are enabled in all "RISC-V modes" at the same time.
  - All PM related CSRs are enabled in just one specific "RISC-V mode" at the time.
  - **Out of Scope**
  - **Comment:** Context switching is not supported in `arch-tests`, so this test is out of scope based on the current state of Sail/ACTs.

**Comment by martinmaas** [comment]
Thanks @UmerShahidengr! I'll let @Adam-pi3 chime in as well, but it makes sense to me that some of the features cannot be supported in the Sail model.

However, I had the same question as Adam:

> Additionally, for the test which we have, I believe we should verify address with and without tag value

The part that was unclear to me is where it checks that the address that is being accessed after the masking operation is the correct one. For example, how do we check whether the implementation masked 7 bits or 16 bits? Or is the idea that we will trap on every access and log the address, thus making the masked address part of the signature?

> It is not checked within the trap handler if PM state is getting modified or not, but after exiting the trap, it is checked in most of the tests if the state persists or not.

I probably overlooked something really simple, but wasn't able to find where this check is happening. Can you please point me to the logic?

**Comment by Adam-pi3** [comment]
Look OK (especially if something is not supported). However, some questions which @martinmaas also pointed out are still open.

**Comment by jjscheel** [comment]
@UmerShahidengr, I believe you have answers for some of @martinmaas questions.  Can you kindly post them for all to see?

**Comment by jjscheel** [comment]
Email from Umer:
> > 1. What's left to complete for tests?  When will we have a PR for tests?

>  I have addressed all comments on the SoW, and in the latest commit at ACT PR, I have completed the following tests:
 All sv48 tests for M/S/U mode with 4 different variants i.e, when mask bits are zero and bit[47]=0 (no trap case for PMM disabled/enabled), mask bits are non-zero and bit[47]=1 (no trap when PMM enabled but trap when disabled due to invalid Virtual Address), when mask bits are zero and bit[47]=1 (no trap when PMM enabled but trap when disabled due to invalid Virtual Address), when mask bits are non-zero and bit[47]=1 (no trap when PMM enabled, may or may not trap for PMM disabled based on if masked bit are all ones or not).
For SV57 tests, we came to know that those are not possible for as sail model does not support sv57 (and spike does not support pmm), satp is not setting its MODE bits to sv57. Although we have integrated the support of pmm in Sail model, and its implementation has been verified via debugging sail code, but its ACTs cant be written because sail is not setting up PTEs for sv57. I added the tests for sv57 in my PR but now I have removed it in my latest commit, since it is not passing the coverpoints and satp is not setting up properly.
For Interrupts and NMIs, ACTs are not possible because of the same constraints on Sail.
In the limited support, these are the only tests which could have been written, and it should be enough proof of concept for ratification

>> 2. Are the coverpoints complete or do we need to extend coverage based on Adam's suggestions?  If so, when will that be done?  When will we have a PR?

>The coverpoints are complete but we have not release those coverpoints yet. The coverpoints require effective address and translated address for comparison in order to check the behavior. Both of these addresses are not provided in the Sail log, coverage tool (RISC-V ISAC) parses the Sail log to check the coverpoints. In order to tackle this issue, RISC-V ISAC tool has been updated, a new architecture state is defined as part of each instruction in the tool which is holding the effective address value and translated address value. This update has been working on our side, but for public release, we need to pass it from an internal review (a company's open source policy), we are internally reviewing those changes, it will be released as PR to RISC-V ISAC. Once RISC-V ISAC PR is added, we will add the coverpoint definitions which will be supported by the tool.
 
>> 3. What is the state of our current test execution?  Have we run them successfully on any simulator?

> Currently all tests are running on riscof, and showing the correct behaviour on Sail model, its signatures are not matching with Spike as spike does not support pmm. but we have verified every test's behaviour by looking at the Sail log, all coverpoints are hitting. Once the coverpoints will be released (within one week), I will let you all know to review the tests against the coverpoints, and I will release the coverage reports.

**Comment by jjscheel** [comment]
Thanks, @UmerShahidengr for the update.  Please post the PR for all at your earlierst convenience so that @Adam-pi3 and @martinmaas can review.  I'll prepare the vote.

**Comment by martinmaas** [comment]
Fantastic – thank you so much for the detailed update! One quick comment on item (3): Pointer masking is now supported in Spike, so it should be possible to compare the two.

https://github.com/riscv-software-src/riscv-isa-sim/pull/1718

**Comment by MuhammadHammad001** [comment]
The tests and covergroups are complete and are available at https://github.com/riscv-non-isa/riscv-arch-test/pull/484
@jjscheel @martinmaas

**Comment by jjscheel** [comment]
@martinmaas, @Adam-pi3, your review of the PR for completeness is what I need for now.  Please post comments here.

**Comment by martinmaas** [comment]
Thank you for all the work on this, and for sending the PR!

The question I would like to confirm is how the following part works:

> In order to tackle this issue, RISC-V ISAC tool has been updated, a new architecture state is defined as part of each instruction in the tool which is holding the effective address value and translated address value. [...] Once RISC-V ISAC PR is added, we will add the coverpoint definitions which will be supported by the tool.

I am not familiar with the coverpoint definition format, but was unable to spot where this was defined.

Assuming that these definitions are present, how are the effective and translated addresses validated? From my understanding, they would need to either be checked explicitly as part of the coverpoint definition, or compared against a run of a different simulator (likely Spike), to confirm they are the same.

Further, have the ACTs been validated against Spike as well, in addition to the Sail model?

Subject to those two points (checking of the effective/translated addresses, and comparison against Spike), the PR looks good to me.

**Comment by jjscheel** [comment]
@UmerShahidengr, @MuhammadHammad001, can we please keep the questions being asked here moving?  We need Adam and Martin to approve your work in the next couple days.

**Comment by jjscheel** [comment]
@Adam-pi3, @martinmaas, with respect to the freeze waiver for Pointer Masking, are you satisfied that we've created the tests for expected coverage?  We may still have debug work and stylistic things to clean up, but we're "feature complete" for testing.

**Comment by martinmaas** [comment]
Thank you for the update! This addresses my concern about comparison against Spike (I am not too concerned if there is a minor bug somewhere, as long as we understand what the bug is and that it does not point towards a problem with the spec; which does not seem to be the case here).

My remaining question is about how the masked addresses are being checked so that they are the same between Sail and Spike – there is a good chance this is somewhere in the logs, but I was not able to see it reading through them.

There is probably an argument to be made that the tests could be considered feature complete even without these comparisons (since the tests themselves already check whether or not masking is applied, just not that the exact right bits are being masked) – but independently of whether this is required for the "feature complete" milestone, it would be good to have this eventually, as it will make coverage significantly stronger.

I'd be interested in @Adam-pi3's view on this as well.

**Comment by UmerShahidengr** [comment]
The masked address in getting checked within the Sail and Spike log, and the cover points are written in such a way that in some of the tests, test is showing different behaviors when pmm is disabled vs pmm is enabled. For example, in one of the tests, we used virtual address as 0xABAB500000000000, this test will cause a trap (due to illegal virtual address) when pmm is disabled, and when pmm will be enabled, then its 0xABAB bits will be masked with zeros and it will not cause trap. Similarly for virtual addresses like 0x0000900000000000, it will cause trap when pmm is disabled, but when pmm is enabled then last bits will be sign extended and masked address will become 0xFFFF900000000000 and will not trap. All these scenerios are tested in the tests.
Also whenever there will be a trap, mtval will get masked address, which is also getting verified via Sail and Spike logs.

**Comment by martinmaas** [comment]
Thanks for the clarification! This is how I read the tests as well, and I agree this tests the most important aspects of pointer masking behavior (i.e., "is it enabled exactly when it should be?").

What the trap-based testing doesn't catch are cases where the implementation masks the wrong bits (e.g., 0x0000900000000000 turns into 0x0010900000000000 despite pointer masking being disabled) – in both cases we'd see a trap, but we don't see that the address was still incorrect. In other words, we get one bit of signal per check, when there are many more incorrect behaviors that cannot be discerned based on that one bit. Having said this, those failure modes don't seem to be like the most likely ones.

I believe there is a good case to be made that the current version should be sufficient in terms of feature-completeness. But if there is a way to test the sequence of actual addresses, I think that would represent a significant increase in coverage.

If everyone else is fine with this, I think it doesn't have to block feature completeness sign-off, though.

**Comment by Adam-pi3** [comment]
Sorry for delay in replying but I was trying to carefully analyze the tests. Per my understanding most of the cover points are covered (the one which are possible to implement taking into account some limitation which @UmerShahidengr was and @MuhammadHammad001 were describing). However, I don't really understand how exactly comparison of the expected addresses in case of fault happens. Tests expects to generate a fault (e.g., when PM is disabled or enabled - depends on the test). However, I don't see the function set-up in MTVEC which would handle this fault and do the expected logic. Is the expectation that tests don't require it and some default fault handler writes something to the log which are parsed? That's the only missing point I have.

I agree we shouldn't block feature completeness sign-off because of that. Nevertheless, my question remains and it is also connected to the @martinmaas comments too.

**Comment by jjscheel** [comment]
Thanks, @Adam-pi3, @martinmaas.  I will mark the Freeze Waiver complete based on your approvals.

Now, the final question becomes have we completed ACT Testing?  This is a Ratification Requirement.

**Comment by martinmaas** [comment]
Thank you @jjscheel! I suggest we try to reach a conclusion on the latter question as well, to avoid delaying ratification.

I think the question that @Adam-pi3 mentioned should be the only remaining item for me as well.

**Comment by jjscheel** [comment]
@UmerShahidengr, can you kindly drive closure on the state of ACT Testing ASAP?

**Comment by UmerShahidengr** [comment]
> Sorry for delay in replying but I was trying to carefully analyze the tests. Per my understanding most of the cover points are covered (the one which are possible to implement taking into account some limitation which [@UmerShahidengr](https://github.com/UmerShahidengr) was and [@MuhammadHammad001](https://github.com/MuhammadHammad001) were describing). However, I don't really understand how exactly comparison of the expected addresses in case of fault happens. Tests expects to generate a fault (e.g., when PM is disabled or enabled - depends on the test). However, I don't see the function set-up in MTVEC which would handle this fault and do the expected logic. Is the expectation that tests don't require it and some default fault handler writes something to the log which are parsed? That's the only missing point I have.
> 

@Adam-pi3 for the trap handler, we have made no changes in the default trap handler, the default trap handler of riscv-arch-tests ([available here](https://github.com/riscv-non-isa/riscv-arch-test/blob/ce04b4930545ae4c81e2f3b6f6935e2aac08679e/riscv-test-suite/env/arch_test.h#L1084C1-L1084C36)) is generic, and dealing with the traps caused by load/store access due to pmm/vm. The default trap handler deposits 4 signature items in the signature file per trap, 2 of them are mepc and mtval value, in case of pmm enabled/disabled, mtval csr will be updated with the effective address, and the same value will be shown in the trap signatures. This value is getting compared in Spike and Sail log

**Comment by UmerShahidengr** [comment]
> What the trap-based testing doesn't catch are cases where the implementation masks the wrong bits (e.g., 0x0000900000000000 turns into 0x0010900000000000 despite pointer masking being disabled) – in both cases we'd see a trap, but we don't see that the address was still incorrect. In other words, we get one bit of signal per check, when there are many more incorrect behaviors that cannot be discerned based on that one bit. Having said this, those failure modes don't seem to be like the most likely ones.

@martinmaas as explained in the above comment, since mtval is getting the effective address on every trap and also getting updated into trap signature area, thus such cases have already been checked if correct bits are getting masked or not. We have made sure to check the effective address on every memory access through the log file, and also made sure to compare mtval value through signature on every trap.

**Comment by jjscheel** [comment]
@martinmaas and @Adam-pi3, does this answer your question?

@UmerShahidengr, so you believe all tests are passing and the coverage is being met?  Correct?

**Comment by martinmaas** [comment]
Thank you for the clarification, Umer! To confirm my understanding: When there is a trap, the effective address will be written into the signature file, and thus compared against a reference implementation. When there is no trap, we know the address is correct, because the correct value is loaded.

Assuming the above, all sounds good to me and I think the tests meet the required coverage.

One final question (but not required for coverage): Was the checking of the effective address in the log file done manually, or can this be done automatically as part of running the test?

**Comment by UmerShahidengr** [comment]
@martinmaas in the developement phase, the effective address was checked from the log file manually, now it is the part of signature file, so one can automatically check the effective address from signature file (in case of a trap).

**Comment by UmerShahidengr** [comment]
@Adam-pi3 , @martinmaas is this PR good to go? Should we merge this one?

**Comment by UmerShahidengr** [comment]
This PR is awaiting the Sail PR to get merged

**Comment by nadime15** [comment]
Hi all,

I have opened a new pull request [#969](https://github.com/riscv/sail-riscv/pull/969) for the Pointer Masking Extension which builds on the work in [#454](https://github.com/riscv/sail-riscv/pull/454) from @HAMZA-AFZAL404 .

I have not had much time yet to thoroughly test the implementation, but I noticed that ACT has already added a set of tests. I created the PR to help move things forward so we can begin using those tests and verify them against the RISC-V Sail model.

I am still getting familiar with ACT, but in the meantime, I would really appreciate it if you could try out the current implementation and share any feedback you might have.

Since Sail does not yet support the Hypervisor, the Hypervisor related part is excluded from this PR. Do not forget to enable the extensions, they are disabled by default.

**Comment by jjscheel** [comment]
@nadime15, here's the ACT for Pointer Masking.  @UmerShahidengr, please work with Nadime to help drive the Sail and ACT work into the communities.

**Comment by jjscheel** [comment]
Work being drive by Certification Test Plan WG and ACT SIG. Details can be found in the [Certification Implementation Tracking](https://docs.google.com/spreadsheets/d/1EuG4LV1TQDx4wTzYDWEAvq0fwaWjNuxenwDARZmrN-E/edit?gid=0#gid=0) spreadsheet.

Closing.
